### ISSUE TITLE
Doesn't play well woth grida-gadget

### ISSUE THREAD

**Issue by falconman44** [vfsfitvnm/frida-il2cpp-bridge#95]
Wow I butchered that title. Eek.

Hello! I recently created a frida-gadget modded APK using this and it doesn't seem to play well at all. I've had 2 users who happen to be using Android 12 tell me that this library causes their graphics to look like [this](https://imgur.com/a/v44ClQw) on load.

I went through basic tests. Extracted an APK and repacked it then told them to run it. No problems.
Injected frida-gadget with an empty script. No problems.
Gave them a script that looked like:

`import "frida-il2cpp-bridge";
Il2Cpp.perform(() => {});`

Graphics looked awful.

I asked them what architecture they're running and all of ours match. The biggest difference I can see is I run Android 10 and they both run 12. This app doesn't want to open in any form of emulator be it Android Studio or Bluestacks so I can't really test it.

Something in this library or a dependency is causing some kind of issue and I don't know what. I figured I'd come here first to see if you have any idea what might be causing this before I explore other avenues.

EDIT: I'm not actually running the latest version. Trying that right now first.

EDIT2: I've gotten them to try a version with the latest frida-il2cpp-bridge and the issue still occurs.

**Comment by falconman44** [comment]
I guess I should've read the other ones here. This sounds similar to #66 except no crashes just garbled graphics. I got the user in question to run a version with the entire perform method commented out and there were no issues. Something in Il2Cpp.perform is causing issues with Android 12.

**Comment by vfsfitvnm** [comment]
Thanks for the investigation.
I suspect this related to #66. Unfortunately I do not own an A12 device, and Frida doesn't work on an A12 emulator (https://github.com/frida/frida/issues/1917)

**Comment by falconman44** [comment]
Well the raspberry Pi 4 has an unofficial android 12 build so I'll try getting that running and replicating the issue then figuring out exactly what may be causing this.

**Comment by vfsfitvnm** [comment]
Thanks, the problem probably occurs here https://github.com/vfsfitvnm/frida-il2cpp-bridge/blob/master/src/utils/native-wait.ts.
On Android, I hook `dlopen` and `android_dlopen_ext` from `libdl.so` to get notified when the il2cpp library is loaded; so a good first test script may be this one:

```ts
const dl = Process.getModuleByName("libdl.so");
Interceptor.attach(dl.getExportByName("dlopen"), {
    onEnter(args) {
        console.log(`dlopen loading ${args[0].readUtf8String()}`);
    }
});
```

**Comment by falconman44** [comment]
So the app freezes on the pi instead of garbled graphics but I can still play around with this.

If I run that script the output is

```
02-15 05:38:13.901  4888  4888 V frida-lief: dlopen loading /data/app/~~KM_bkKS16D6oJVMNpHRh2A==/com.spaceapegames.beatstar-kvXWBVE4QaOeHhlt1i1_tg==/lib/arm64/libunity.so
02-15 05:38:13.996  4888  4888 V frida-lief: dlopen loading /vendor/lib64/gbm/v3d_gbm.so
02-15 05:38:13.999  4888  4888 V frida-lief: dlopen loading libglapi.so.0
02-15 05:38:14.002  4888  4888 V frida-lief: dlopen loading /vendor/lib64/dri/v3d_dri.so
02-15 05:38:14.004  4888  4888 V frida-lief: dlopen loading libglapi.so.0
02-15 05:38:14.007  4888  4888 V frida-lief: dlopen loading /vendor/lib64/dri/kms_swrast_dri.so
02-15 05:38:14.008  4888  4888 V frida-lief: dlopen loading libglapi.so.0
02-15 05:38:14.011  4888  4888 V frida-lief: dlopen loading /vendor/lib64/dri/swrast_dri.so
```

(I've adapted the script [here](https://lief-project.github.io/doc/stable/tutorials/09_frida_lief.html) to print it out as a gadget script)

Commenting out https://github.com/vfsfitvnm/frida-il2cpp-bridge/blob/master/src/il2cpp/base.ts#L117 eliminates the crash so it does look like it originates from native-wait somewhere.

**Comment by vfsfitvnm** [comment]
Hmm, does the minimal snippet I provided still freeze the process?

**Comment by falconman44** [comment]
Yes actually, it does.

```
const dl = Process.getModuleByName("libdl.so");
Interceptor.attach(dl.getExportByName("dlopen"), {
    onEnter(args) {}
});
```

Is enough to crash it.

**Comment by vfsfitvnm** [comment]
That's very bad. What about `android_dlopen_ext` instead of `dlopen`?

**Comment by falconman44** [comment]
No crash there.

```
const dl = Process.getModuleByName("libdl.so");
Interceptor.attach(dl.getExportByName("android_dlopen_ext"), {
    onEnter(args) {
		  var Log = Java.use("android.util.Log");
		  Log.v("frida-lief", `dlopen loading ${args[0].readUtf8String()}`);
    }
});

02-15 06:06:43.411  7695  7695 V frida-lief: dlopen loading /data/app/~~KM_bkKS16D6oJVMNpHRh2A==/com.spaceapegames.beatstar-kvXWBVE4QaOeHhlt1i1_tg==/lib/arm64/libmain.so
02-15 06:06:43.513  7695  7695 V frida-lief: dlopen loading /vendor/lib64/hw/android.hardware.graphics.mapper@4.0-impl.minigbm_gbm_mesa.so
02-15 06:06:44.296  7695  7765 V frida-lief: dlopen loading vulkan.broadcom.so
02-15 06:06:44.316  7695  7765 V frida-lief: dlopen loading /vendor/lib64/hw/gralloc.default.so
02-15 06:06:44.338  7695  7765 V frida-lief: dlopen loading /vendor/lib64/hw/android.hardware.graphics.mapper@4.0-impl.minigbm_gbm_mesa.so
02-15 06:06:45.226  7695  7765 V frida-lief: dlopen loading /data/app/~~KM_bkKS16D6oJVMNpHRh2A==/com.spaceapegames.beatstar-kvXWBVE4QaOeHhlt1i1_tg==/lib/arm64/libAkSoundEngine.so
```

**Comment by vfsfitvnm** [comment]
Ok, so `dlopen` cannot be touched on A12 - `Interceptor.attach` actually patches the instruction prelude so it gets redirected somewhere else. I included `dlopen` because some applications uses that one instead of `android_dlopen_ext` when it comes to load `libil2cpp.so`.
So we should probably hook a lower level function inside `linker(64)`...

**Comment by vfsfitvnm** [comment]
Infact `com.spaceapegames.beatstar` loads `libil2cpp.so` via `dlopen` on my A10 device

**Comment by vfsfitvnm** [comment]
Would you try:
```ts
Interceptor.attach(Module.getExportByName(null, "__loader_dlopen"), {
    onEnter(args) {
        console.log(`__loader_dlopen loading ${args[0].readUtf8String()}`);
    }
});
```

**Comment by falconman44** [comment]
```
02-15 06:21:44.626  8401  8401 V frida-lief: __loader_dlopen loading /data/app/~~KM_bkKS16D6oJVMNpHRh2A==/com.spaceapegames.beatstar-kvXWBVE4QaOeHhlt1i1_tg==/lib/arm64/libunity.so
02-15 06:21:44.731  8401  8401 V frida-lief: __loader_dlopen loading /vendor/lib64/gbm/v3d_gbm.so
02-15 06:21:44.735  8401  8401 V frida-lief: __loader_dlopen loading libglapi.so.0
02-15 06:21:44.736  8401  8401 V frida-lief: __loader_dlopen loading /vendor/lib64/dri/v3d_dri.so
02-15 06:21:44.804  8401  8469 V frida-lief: __loader_dlopen loading libc.so
02-15 06:21:44.917  8401  8469 V frida-lief: __loader_dlopen loading /data/app/~~KM_bkKS16D6oJVMNpHRh2A==/com.spaceapegames.beatstar-kvXWBVE4QaOeHhlt1i1_tg==/lib/arm64/libil2cpp.so
02-15 06:21:45.453  8401  8469 V frida-lief: __loader_dlopen loading libvulkan.so
02-15 06:21:45.477  8401  8469 V frida-lief: __loader_dlopen loading /vendor/lib64/dri/tls/v3d_dri.so
02-15 06:21:45.478  8401  8469 V frida-lief: __loader_dlopen loading /vendor/lib64/dri/v3d_dri.so
02-15 06:21:45.531  8401  8469 V frida-lief: __loader_dlopen loading libGLESv2.so
02-15 06:21:46.369  8401  8469 V frida-lief: __loader_dlopen loading /data/app/~~KM_bkKS16D6oJVMNpHRh2A==/com.spaceapegames.beatstar-kvXWBVE4QaOeHhlt1i1_tg==/lib/arm64/libAkSoundEngine.so
02-15 06:21:46.529  8401  8469 V frida-lief: __loader_dlopen loading libaaudio.so
```

**Comment by vfsfitvnm** [comment]
Did it crash or whatever?

**Comment by falconman44** [comment]
No crashes. Should've mentioned that when I posted oops.

**Comment by vfsfitvnm** [comment]
Ok, that's good (`dlopen` calls `__loader_dlopen`, we aren't missing anything); so I can say we *could* have a fix. However, I'd like to know if we can completely remove `android_dlopen_ext` at this point, none of the apps I tested actually uses it...

Meanwhile, would you replace [this](https://github.com/vfsfitvnm/frida-il2cpp-bridge/blob/master/src/utils/native-wait.ts#L19) line with:
```ts
return [null, ["__loader_dlopen", "utf8"]];
```
and see if `Il2Cpp::perform` works?

**Comment by falconman44** [comment]
Once replaced and a small change made, yes it does work.

https://github.com/vfsfitvnm/frida-il2cpp-bridge/blob/master/src/il2cpp/base.ts#L125

When running as a frida-gadget script at least, perform doesn't run unless you add a call to resolve() right there but no crashing and Il2Cpp.perform has run.

I can't exactly test my script though because the game requires two fingers and I can't replicate that on a raspberry pi. I'll try passing it on to the user reporting issues and see if they can use it,

**Comment by vfsfitvnm** [comment]
Hmm, `setImmediate(resolve);` should be equivalent to 
```ts
setImmediate(() => resolve());
```
Did I understand correctly?

**Comment by falconman44** [comment]
I thought they should be equivalent too. This behaviour only seems to occur when running it with frida-gadget as an autorun script though so I've just been adding a resolve call manually when building. It works fine without frida-gadget.

**Comment by vfsfitvnm** [comment]
But did you wrap the `resolve` call inside a `setImmediate`?

**Comment by falconman44** [comment]
I've tried lots of methods. In summary:

```
Il2Cpp.perform(() => {
    var Log = Java.use("android.util.Log");
    Log.v("frida-lief", "Fixed");
});
```

```
if (Il2Cpp.Api._getCorlib().isNull()) {
    await new Promise(resolve => {
        const interceptor = Interceptor.attach(Il2Cpp.Api._init, {
            onLeave() {
                interceptor.detach();
                setImmediate(() => resolve()); //function call
            }
        });
    });
}
```

Prints nothing

```
if (Il2Cpp.Api._getCorlib().isNull()) {
    await new Promise(resolve => {
        const interceptor = Interceptor.attach(Il2Cpp.Api._init, {
            onLeave() {
                interceptor.detach();
                setImmediate(() => resolve); //not function call
            }
        });
    });
}
```

Prints nothing

```
if (Il2Cpp.Api._getCorlib().isNull()) {
    await new Promise(resolve => {
        const interceptor = Interceptor.attach(Il2Cpp.Api._init, {
            onLeave() {
                interceptor.detach();
                setImmediate(() => resolve);
                resolve();
            }
        });
    });
}
```

Prints 

`02-15 06:42:26.716 10452 10523 V frida-lief: Fixed`

This isn't an Android 12 issue. It happens with my physical Android 10 device as well.

**Comment by vfsfitvnm** [comment]
Hmm so there's another issue here. We must use `setImmediate` to let our code execute inside frida's thread. In this way:
```ts
await new Promise(resolve => {
        const interceptor = Interceptor.attach(Il2Cpp.Api._init, {
            onLeave() {
                interceptor.detach();
                resolve();
            }
        });
    });
```
Our code will keep running inside the il2cpp main thread and this is proven to break many applications

**Comment by falconman44** [comment]
Well I gave the user experiencing issues a version with the `return [null, ["__loader_dlopen", "utf8"]];` replacement and my awful resolve() fix and they're reporting that it works now. No graphical issues and the script runs successfully.

**Comment by vfsfitvnm** [comment]
@falconman44 I'm trying gadget too, however in which library did you inject gadget dependency?

**Comment by falconman44** [comment]
I used objection to automate it: https://github.com/sensepost/objection

`objection patchapk -s '.\Beatstar - Touch Your Music_17.0.0.18644_apkcombo.com.apk' -c .\config.json -l .\index.js`

Where config.json is

```
{
  "interaction": {
    "type": "script",
    "path": "libfrida-gadget.script.so"
  }
}
```

It's output is:

```
No architecture specified. Determining it using `adb`...
Detected target device architecture as: arm64-v8a
Using latest Github gadget version: 15.1.17
Patcher will be using Gadget version: 15.1.17
Detected apktool version as: 2.6.0
Running apktool empty-framework-dir...
I: Removing 1.apk framework file...
Unpacking .\Beatstar - Touch Your Music_17.0.0.18644_apkcombo.com.apk
App already has android.permission.INTERNET
Target class not specified, searching for launchable activity instead...
Smali not found in smali directory. This might be a multidex APK. Searching...
Found smali at: C:\Users\megam\AppData\Local\Temp\tmpln5e9ayu.apktemp\smali_classes3\com/unity3d/player/UnityPlayerActivity.smali
Reading smali from: C:\Users\megam\AppData\Local\Temp\tmpln5e9ayu.apktemp\smali_classes3\com/unity3d/player/UnityPlayerActivity.smali
Injecting loadLibrary call at line: 13
Attempting to fix the constructors .locals count
Current locals value is 0, updating to 1:
Writing patched smali back to: C:\Users\megam\AppData\Local\Temp\tmpln5e9ayu.apktemp\smali_classes3\com/unity3d/player/UnityPlayerActivity.smali
Copying Frida gadget to libs path...
Adding a gadget configuration file...
Copying over a custom script to use with the gadget config.
Rebuilding the APK with the frida-gadget loaded...
Built new APK with injected loadLibrary and frida-gadget
Performing zipalign
Zipalign completed
Signing new APK.
Signed the new APK
```

**Comment by vfsfitvnm** [comment]
Thanks, I could reproduce

**Comment by vfsfitvnm** [comment]
Minimal snippet to reproduce:
```ts
const Log = Java.use("android.util.Log");

Log.e("frida-log", "Start");

new Promise<void>(resolve => setTimeout(resolve, 1000)).then(() => Log.e("frida-log", `Ready!`));
```
This looks like a frida bug.

**Comment by vfsfitvnm** [comment]
@falconman44 
The reason why it doesn't work is because you are not attached to a java vm when trying to print.
```ts
new Promise<void>(resolve => setTimeout(resolve, 1000)).then(() => {
    Java.use("android.util.Log").e("frida-log", `Ready!`);
});
```
Instead, the following works:
```ts
new Promise<void>(resolve => setTimeout(resolve, 1000)).then(() => {
    Java.perform(() => Java.use("android.util.Log").e("frida-log", `Ready!`));
});
```

However, for some reason, this only happens when `setTimeout` or `setImmediate` are involved.

I guess this can be closed as soon as I provide A12 support (very soon)!

**Comment by vfsfitvnm** [comment]
Fixed in 0.7.5. Thanks @falconman44 !
